Terraform AppEngine Dispatch Resource

callppatel opened the pull request:
<!-- AUTOCHANGELOG for Downstream PRs.

EXTERNAL CONTRIBUTORS: Ignore please - your reviewer will handle.

INTERNAL CONTRIBUTORS AND REVIEWERS: See .ci/RELEASE_NOTES_GUIDE.md
for writing good release notes.

NO CHANGELOG NOTE: Please add "changelog: no-release-note" label to this PR.

Otherwise, fill the template out (replace the heading).
You can add more release notes if you want more than one CHANGELOG entry for
this PR, but make sure not to indent notes and to leave newlines between
code blocks for Markdown's sake.

For Terraform PRs, we use the following "release-note:" headings
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
-->
This is follow-up of https://github.com/GoogleCloudPlatform/magic-modules/pull/2338
Created a new pull request as there was some conflict in repository

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_app_engine_application_url_dispatch_rules`
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.  They will authorize it to run through our CI pipeline, which will generate downstream PRs.

Thanks for your contribution!  A human will be with you soon.

@emilymye, please review this PR or find an appropriate assignee.

callppatel commented:
@rileykarson - Please review this, as earlier pull request had some conflict.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 9a74b02e2bff552ebdacc4c55cc97072886378db. 
## Pull request statuses
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/1262 
depends: https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/227 
depends: https://github.com/terraform-providers/terraform-provider-google/pull/4674

rileykarson commented on third_party/terraform/website-compiled/google.erb:
```diff
@@ -235,6 +235,9 @@
       <a href="/docs/providers/google/r/app_engine_standard_app_version.html">google_app_engine_standard_app_version</a>
       </li>
+      <li<%%= sidebar_current("docs-google_app_engine_application_url_dispatch_rules") %>>
+      <a href="/docs/providers/google/r/google_app_engine_application_url_dispatch_rules.html">google_app_engine_application_url_dispatch_rules</a>
```
```suggestion
      <a href="/docs/providers/google/r/app_engine_application_url_dispatch_rules.html">google_app_engine_application_url_dispatch_rules</a>
```

rileykarson commented on third_party/terraform/website-compiled/google.erb:
```diff
@@ -235,6 +235,9 @@
       <li<%%= sidebar_current("docs-google-app-engine-standard-app-version") %>>
       <a href="/docs/providers/google/r/app_engine_standard_app_version.html">google_app_engine_standard_app_version</a>
       </li>
+      <li<%%= sidebar_current("docs-google_app_engine_application_url_dispatch_rules") %>>
```
```suggestion
      <li<%%= sidebar_current("docs-google-app-engine-application-url-dispatch-rules") %>>
```

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -446,3 +446,42 @@ objects:
           AutomaticScaling F1, F2, F4, F4_1G
           (Only AutomaticScaling is supported at the moment)
+  - !ruby/object:Api::Resource
+    name: 'ApplicationUrlDispatchRules'
```
We need to add an `async` block so that Terraform knows to wait on async operations from the API; https://github.com/GoogleCloudPlatform/magic-modules/blob/master/products/appengine/api.yaml#L41-L57 contains an example.

I receive this error otherwise:

```
---[ RESPONSE ]--------------------------------------
HTTP/2.0 409 Conflict
Alt-Svc: quic=":443"; ma=2592000; v="46,43",h3-Q048=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Mon, 14 Oct 2019 19:47:55 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "error": {
    "code": 409,
    "message": "Cannot operate on apps/my-project because an operation is already in progress for apps/my-project by 9152be65-8e22-417e-a0d7-3c4151befddb.",
    "status": "ABORTED"
  }
}

```

callppatel replied:
Could you please review the async block. I really doesnt have idea about path & base_url.

rileykarson replied:
It should be good, they're generally common inside products (and most are common across most products). Generating now to make sure.

callppatel replied:
Thank you.  I was able to test this locally. Please let me know if your test is working. Thanks for your help !

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, def768f573a6e9f9e1150f408b697d1115c21982. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

callppatel commented:
@rileykarson - Can we work on this today ? Please let me know if any open feedback. Thanks.

callppatel commented:
StandardAppVersion already have mutex, added this for ApplicationUrlDispatchRules. Please check if your test is working. 

Are you getting this error when you are try to create new service and apply the Dispatch URL in same terraform code?

callppatel commented:
changed Mutex and also added depends_on = ["google_app_engine_standard_app_version.myapp_v1"] in templates/terraform/examples/app_engine_application_url_dispatch_rules_basic.tf.erb

Looks good at my end.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 46928a17fdd5c4a4e00231abc56c3aa7d74ec752. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

callppatel commented:
> It looks like this is failing a test because the resource can't be deleted, and Terraform sees that it still exists. You can add a `custom_check_destroy` that's empty to check for it getting deleted: https://github.com/GoogleCloudPlatform/magic-modules/blob/master/provider/terraform/custom_code.rb#L107-L110.
> 
> You can reuse https://github.com/GoogleCloudPlatform/magic-modules/blob/master/templates/terraform/custom_check_destroy/appengine_version.go.erb, it's probably best if we also rename it to `appengine.go.erb` since it's being used by multiple resources.

Added appengine.go.erb as per above recommendation

rileykarson commented on products/appengine/terraform.yaml:
```diff
@@ -93,6 +93,19 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+    import_format: ["{{project}}"]
+    mutex: "apps/{{project}}"
+    custom_code: !ruby/object:Provider::Terraform::CustomCode
+      custom_delete: templates/terraform/custom_delete/appengine.go.erb
```
```suggestion
```

callppatel replied:
I have added the same file at same location: templates/terraform/custom_delete/appengine.go.erb

rileykarson replied:
Ah, you're right. Unlike the `test_check_destroy`, that file is specific to `StandardAppVersion`. Can you delete it and remove the reference?

callppatel replied:
ya, I replaced with ApplicationUrlDispatchRules. Not sure rest of the code in same file is correct or not.. just copied from other file.

rileykarson replied:
We don't want the custom delete function, and if we add one it will be different- I believe deletion was working correctly before.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, b44e9411454088b1c235360fb7efac7c1b065297. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on products/appengine/terraform.yaml:
```diff
@@ -93,6 +93,18 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+      test_check_destroy: templates/terraform/custom_check_destroy/appengine.go.erb
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "app_engine_application_url_dispatch_rules_basic"
```
```suggestion
        name: "app_engine_application_url_dispatch_rules_basic"
        primary_resource_id: "service_rules"
```

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 1a133a09720193c728768f3134bad4361e3a79c7. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/ansible/ansible_collections_google/pull/30

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -446,3 +446,59 @@ objects:
+    properties:
+      - !ruby/object:Api::Type::Array   
+        name: 'dispatchRules'
+        description: |
```
```suggestion
        required: true
        description: |
```

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -446,3 +446,59 @@ objects:
+              description: |
+                Domain name to match against. The wildcard "*" is supported if specified before a period: "*.".
+                Defaults to matching all domains: "*".
+              required: true
```
```suggestion
```

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -446,3 +446,59 @@ objects:
+        item_type: !ruby/object:Api::Type::NestedObject
+          properties:
+            - !ruby/object:Api::Type::String
+              name: 'domain'
```
```suggestion
              name: 'domain'
              default_value: '*'
```

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -446,3 +446,59 @@ objects:
+                Pathname within the host. Must start with a "/". A single "*" can be included at the end of the path.
+                The sum of the lengths of the domain and path may not exceed 100 characters.
+              required: true
+              
```
```suggestion
```

rileykarson commented on products/appengine/terraform.yaml:
```diff
@@ -93,6 +93,19 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+    import_format: ["{{project}}"]
+    mutex: "apps/{{project}}"
+    custom_code: !ruby/object:Provider::Terraform::CustomCode
+      test_check_destroy: templates/terraform/custom_check_destroy/appengine.go.erb
```
Can you remove the old `custom_check_destroy` file and use the same one for both appengine resources?

callppatel replied:
Its done

rileykarson commented on templates/terraform/examples/app_engine_application_url_dispatch_rules_basic.tf.erb:
```diff
@@ -0,0 +1,44 @@
+    path = "/myapp/*"
+    service = "myapp"
+  }
+  depends_on = ["google_app_engine_standard_app_version.myapp_v1"]
```
```suggestion
```

rileykarson commented on templates/terraform/examples/app_engine_application_url_dispatch_rules_basic.tf.erb:
```diff
@@ -0,0 +1,44 @@
+  dispatch_rules {
+    domain = "*"
+    path = "/myapp/*"
+    service = "myapp"
```
```suggestion
    service = "${google_app_engine_standard_app_version.myapp_v1.service}"
```

This should obviate the `depends_on`.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 72012cc725c81368873b7314cc7626e4cd2e1b26. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 32bc4d7571eae78b8b7b7e69e8f03d8500045fa2. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 182ccbe712cf7446684e86e7fde155a0ca92e872. 
## Pull request statuses
No diff detected in terraform-provider-google-beta.
No diff detected in terraform-google-conversion.
No diff detected in terraform-provider-google.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.
